### PR TITLE
Fix regression in AssumeRole session policy handling (fixes #12756)

### LINKED ISSUE #12756: STS Permissions in Assume Role Started Giving 403 Errors In RELEASE.2021-07-12T02-44-53Z
The July 15, 2021 update of Minio broke something related to custom policies in Assume Role STS requests.

## Expected Behavior
Custom policies apply to assumed roles.

## Current Behavior
The user of the assumed role gets a 403 "Access Denied" response.

## Possible Solution
After debugging, I have no clue why the access is denied, but I tied it specifically to version minio.RELEASE.2021-07-15T22-27-34Z because reverting to .RELEASE.2021-07-12T02-44-53Z fixed the problem. 

## Steps to Reproduce (for bugs)

1. Create an assumed role with the following parameters:
```
const policy = {
	Version: '2012-10-17',
	Statement: [
		{
			Sid: Math.random().toString().substr(2, 8),
			Action: ['s3:AbortMultipartUpload', 's3:ListMultipartUploadParts', 's3:PutObject'],
			Effect: 'Allow',
			Resource: `arn:aws:s3:::*`,
		},
	],
};

const params: AWS.STS.AssumeRoleRequest = {
	RoleArn: config.fileStorage.ROLE_ARN,
	RoleSessionName: `uploader${Math.random()}`,
	Policy: JSON.stringify(policy),
	DurationSeconds: 3600,
};
const { Credentials } = await sts.assumeRole(params).promise();
return {
	accessKeyId: Credentials.AccessKeyId,
	secretAccessKey: Credentials.SecretAccessKey,
	sessionToken: Credentials.SessionToken,
	expiration: Credentials.Expiration,
};
```
2. Use the resulting creds to upload a file:
```
const testKey = 'test/abc123';
const putParams: AWS.S3.PutObjectRequest = {
	Bucket: fileStorage.bucket,
	Body: 'someData',
	Key: testKey,
};
const credentials = await getTemporaryUploadCreds(testKey);

/** An S3 authenticated with the creds for STS */
const s3 = new AWS.S3({
	...fileStorage.getConfig(),
	credentials,
});

const res = await s3.putObject(putParams).promise();

```

**NOTE: The request will work if no policy is set at all in the params, as it will default to the user's original policy**

## Context

This is used to provide clients with the ability to upload multi-part files to a specific key(s) without requiring the server to sign keys for every part.

One of the PRs in minio.RELEASE.2021-07-15T22-27-34Z caused this issue. Most likely this one related to STS changes: https://github.com/minio/minio/pull/12716/files

## Your Environment

* Version used (`minio --version`): minio.RELEASE.2021-07-15T22-27-34Z
* Server setup and configuration: Minio setup as per Laravel Homestead: https://github.com/laravel/homestead/blob/main/scripts/features/minio.sh
With the following policy setup:
```
# Create minio access group, create new user then add to group.
mc admin user add homestead homestead_app secretkey
mc admin group add homestead rwUsers homestead_app
mc admin policy set homestead readwrite group=rwUsers
```
* Operating System and version (`uname -a`): #66~20.04.1-Ubuntu SMP Thu Jun 17 11:14:10 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

### PR DESCRIPTION
**Pull request by donatello**
## How to test this PR?

Use assume role to create temp creds for a user (see `docs/sts/assume-role.md`):

```
aws --profile foobar --endpoint-url http://localhost:9000 sts assume-role --policy '{"Version":"2012-10-17","Statement":[{"Sid":"Stmt1","Effect":"Allow","Action":"s3:*","Resource":"arn:aws:s3:::*"}]}' --role-arn arn:xxx:xxx:xxx:xxxx --role-session-name anything
```

Check that generated creds works:

```
MC_HOST_ftmp=http://access_key:secret_key:session_token@localhost:9000 mc ls ftmp
```

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] Fixes a regression (introduced in eae9c2f65bade4d9de2aac42f188114bf6dbe555)
- [ ] Documentation updated
- [ ] Unit tests added/updated

### CODE REVIEW

**Review by harshavardhana — APPROVED** [review]

File: cmd/iam.go [hunk 1]
```diff
@@ -2171,36 +2171,14 @@ func (sys *IAMSys) IsAllowedServiceAccount(args iampolicy.Args, parentUser strin
...
-	if err != nil {
-		// Log any error in input session policy config.
-		logger.LogIf(GlobalContext, err)
-		return false
-	}
-
-	// This can only happen if policy was set but with an empty JSON.
-	if subPolicy.Version == "" && len(subPolicy.Statements) == 0 {
-		return combinedPolicy.IsAllowed(parentArgs)
-	}
```
Comment (harshavardhana) [inline cmd/iam.go]: We need this code to avoid a console UI bug where it sends an empty JSON `{}` @donatello

Comment (harshavardhana) [reply cmd/iam.go]: You do not have to make this unification as part of this PR and perhaps revert it if not needed..

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

File: cmd/iam.go (hunk 1 above)
Comment (donatello) [reply cmd/iam.go]: Ok, will remove this part from this PR

**Review by harshavardhana — APPROVED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by harshavardhana** [comment]
PTAL @vadmeste

File: cmd/auth-handler.go [hunk 2]
```diff
@@ -244,6 +234,19 @@ func getClaimsFromToken(token string) (map[string]interface{}, error) {
 		claims.MapClaims[iampolicy.SessionPolicyName] = string(spBytes)
 	}
 
+	// If LDAP claim key is set, return here.
+	if _, ok := claims.MapClaims[ldapUser]; ok {
```
Comment (vadmeste) [inline cmd/auth-handler.go:238]: Can STS LDAP credentials have session policy ? if not let's move this before line 221 so the logic won't be overlapped

Comment (harshavardhana) [reply cmd/auth-handler.go:238]: Yes LDAP can have session policy

**Review by vadmeste — APPROVED** [review]
LGTM

**Pull request merged**